File: backend/services/taskService.js
/**
 * Task Service
 * Orchestrates task lifecycle: creation, scheduling, execution simulation,
 * completion, and SLA tracking. Acts as the central coordinator between
 * scheduling algorithms and the resource manager.
 */

const Task = require('../models/Task');
const resourceManager = require('./resourceManager');
const { greedySchedule } = require('../schedulers/greedy');
const { knapsackSchedule } = require('../schedulers/knapsack');
const { fcfsSchedule } = require('../schedulers/fcfs');
const { prioritySchedule } = require('../schedulers/priority');
const { roundRobinSchedule } = require('../schedulers/roundRobin');

// All tasks in system (active + recent history)
const tasks = new Map();          // id -> Task
const queue = [];                 // Tasks waiting to be scheduled
let currentAlgorithm = 'greedy'; // Default scheduling algorithm

// Valid algorithms
const ALGORITHMS = ['greedy', 'knapsack', 'fcfs', 'priority', 'roundrobin'];

// Priority queue comparator (higher urgency = lower index)
function queueComparator(a, b) {
  return b.getUrgencyScore() - a.getUrgencyScore();
}

function sortQueue() {
  queue.sort(queueComparator);
}

// ─── Algorithm Selection ────────────────────────────────────────────────────

function setAlgorithm(algo) {
  const normalized = algo.toLowerCase().replace(/[-_\s]/g, '');
  if (!ALGORITHMS.includes(normalized)) {
    throw new Error(`Unknown algorithm "${algo}". Valid: ${ALGORITHMS.join(', ')}`);
  }
  currentAlgorithm = normalized;
  return currentAlgorithm;
}

function getAlgorithm() {
  return currentAlgorithm;
}

// ─── Scheduling ─────────────────────────────────────────────────────────────

function scheduleTask(task) {
  const servers = resourceManager.getServers();
  const allTasks = getAllTasks();
  let result = null;

  switch (currentAlgorithm) {
    case 'greedy':
      result = greedySchedule(task, servers);
      break;
    case 'knapsack':
      result = knapsackSchedule(task, servers);
      break;
    case 'fcfs':
      result = fcfsSchedule(task, servers);
      break;
    case 'priority':
      result = prioritySchedule(task, servers, allTasks);
      break;
    case 'roundrobin':
      result = roundRobinSchedule(task, servers);
      break;
    default:
      result = greedySchedule(task, servers);
  }

  if (!result) return false;

  const { server, log, preempted } = result;

  // Handle preemption: remove evicted task from server
  if (preempted) {
    preempted.status = 'preempted';
    preempted.completedAt = new Date();
    resourceManager.deallocate(preempted);
    resourceManager.logDecision(`PREEMPTED Task#${preempted.id} to make room for Task#${task.id}`);
    // Re-queue the preempted task
    preempted.status = 'queued';
    preempted.assignedServer = null;
    preempted.scheduledAt = null;
    preempted.startedAt = null;
    preempted.completedAt = null;
    queue.push(preempted);
    sortQueue();
  }

  // Allocate resources
  const allocated = resourceManager.allocate(task, server);
  if (!allocated) return false;

  // Update task state
  task.assignedServer = server.id;
  task.scheduledBy = currentAlgorithm;
  task.scheduledAt = new Date();
  task.startedAt = new Date();
  task.status = 'running';
  task.waitTime = task.startedAt - task.createdAt;
  task.decisionLog = log;

  resourceManager.logDecision(log);

  // Simulate task execution: complete after executionTime seconds
  const execMs = task.executionTime * 1000;
  setTimeout(() => completeTask(task.id), execMs);

  return true;
}

// ─── Task Creation ──────────────────────────────────────────────────────────

function createTask(params) {
  // Validate required fields
  if (!params.cpu || !params.memory || !params.executionTime) {
    throw new Error('Task requires cpu, memory, and executionTime fields');
  }
  if (params.cpu <= 0 || params.cpu > 100) {
    throw new Error('cpu must be between 1 and 100');
  }
  if (params.memory <= 0 || params.memory > 65536) {
    throw new Error('memory must be between 1 and 65536 MB');
  }
  if (params.executionTime <= 0 || params.executionTime > 3600) {
    throw new Error('executionTime must be between 1 and 3600 seconds');
  }

  const task = new Task(params);
  tasks.set(task.id, task);
  queue.push(task);
  sortQueue();

  // Try to schedule immediately
  const scheduled = tryScheduleFromQueue();
  if (!scheduled) {
    resourceManager.logDecision(
      `Task#${task.id} queued (no resources available, queue depth: ${queue.length})`
    );
  }

  return task;
}

/**
 * Create multiple tasks as a batch
 */
function createBatch(taskParamsList, batchId = null) {
  const bid = batchId || `batch-${Date.now()}`;
  const created = taskParamsList.map(params => {
    const task = new Task({ ...params, batchId: bid });
    tasks.set(task.id, task);
    queue.push(task);
    return task;
  });
  sortQueue();
  // Schedule as many as possible
  let scheduled = 0;
  while (queue.length > 0) {
    const ok = tryScheduleFromQueue();
    if (!ok) break;
    scheduled++;
  }
  resourceManager.logDecision(
    `Batch ${bid}: created ${created.length} tasks, scheduled ${scheduled} immediately`
  );
  return { batchId: bid, tasks: created, scheduled };
}

// ─── Task State Management ───────────────────────────────────────────────────

function completeTask(taskId) {
  const task = tasks.get(taskId);
  if (!task || task.status !== 'running') return false;

  task.status = 'completed';
  task.completedAt = new Date();

  // Check SLA: did it meet its deadline?
  if (task.deadline && task.completedAt > task.deadline) {
    task.slaViolated = true;
    resourceManager.metrics.slaViolations++;
    resourceManager.logDecision(
      `SLA VIOLATION: Task#${task.id} completed at ${task.completedAt.toISOString()} ` +
      `but deadline was ${task.deadline.toISOString()}`
    );
  }

  resourceManager.metrics.totalTasksCompleted++;
  resourceManager.deallocate(task);

  // Try to schedule queued tasks now that resources are free
  tryScheduleFromQueue();

  return true;
}

function cancelTask(taskId) {
  const task = tasks.get(taskId);
  if (!task) return null;

  const prevStatus = task.status;
  if (prevStatus === 'completed' || prevStatus === 'cancelled') return task;

  // Remove from queue if queued
  const qIdx = queue.findIndex(t => t.id === task.id);
  if (qIdx !== -1) queue.splice(qIdx, 1);

  // Deallocate if running
  if (prevStatus === 'running') {
    resourceManager.deallocate(task);
    resourceManager.metrics.totalTasksFailed++;
    tryScheduleFromQueue();
  }

  task.status = 'cancelled';
  task.completedAt = new Date();
  resourceManager.metrics.totalTasksCancelled++;
  resourceManager.logDecision(`Task#${task.id} cancelled (was: ${prevStatus})`);

  return task;
}

// ─── Queue Processing ────────────────────────────────────────────────────────

function tryScheduleFromQueue() {
  if (queue.length === 0) return false;

  // Check for deadline-expired tasks first
  queue.forEach(t => {
    if (t.isDeadlineExpired() && t.status === 'queued') {
      t.slaViolated = true;
      resourceManager.metrics.slaViolations++;
    }
  });

  const task = queue[0]; // Highest urgency task
  const scheduled = scheduleTask(task);
  if (scheduled) {
    queue.shift(); // Remove from queue
    return true;
  }
  return false;
}

// ─── Query ───────────────────────────────────────────────────────────────────

function getTask(id) {
  return tasks.get(id) || null;
}

function getAllTasks() {
  return Array.from(tasks.values());
}

function getQueue() {
  return [...queue];
}

function getTasksByStatus(status) {
  return getAllTasks().filter(t => t.status === status);
}

function getTasksByUser(userId) {
  return getAllTasks().filter(t => t.userId === userId);
}

function getAlgorithmStats() {
  const all = getAllTasks().filter(t => t.scheduledBy);
  const stats = {};
  for (const algo of ALGORITHMS) {
    const byAlgo = all.filter(t => t.scheduledBy === algo);
    const completed = byAlgo.filter(t => t.status === 'completed');
    stats[algo] = {
      scheduled: byAlgo.length,
      completed: completed.length,
      avgWaitTime: completed.length
        ? Math.round(completed.reduce((s, t) => s + t.waitTime, 0) / completed.length)
        : 0,
      slaViolations: byAlgo.filter(t => t.slaViolated).length,
    };
  }
  return stats;
}

module.exports = {
  createTask,
  createBatch,
  cancelTask,
  completeTask,
  getTask,
  getAllTasks,
  getQueue,
  getTasksByStatus,
  getTasksByUser,
  setAlgorithm,
  getAlgorithm,
  getAlgorithmStats,
  ALGORITHMS,
};

File: backend/services/task_service.py
from models.task import Task, TaskStatus
from models.server import Server, ServerStatus
from services.scheduler import SchedulerService
from services.resource_manager import ResourceManager


class TaskService:
    def __init__(self):
        self._tasks: dict[str, Task] = {}
        self._servers: dict[str, Server] = {}
        self._scheduler = SchedulerService([], algorithm="greedy")
        self._resource_manager = ResourceManager()

    # ── Task CRUD ──────────────────────────────────────────────────────────────

    def create_task(self, data: dict) -> Task:
        required = ["name", "cpu_required", "memory_required", "priority"]
        for field in required:
            if field not in data:
                raise ValueError(f"Missing required field: {field}")
        priority = int(data["priority"])
        if not 1 <= priority <= 10:
            raise ValueError("priority must be between 1 and 10")
        cpu = float(data["cpu_required"])
        mem = float(data["memory_required"])
        if cpu <= 0 or mem <= 0:
            raise ValueError("cpu_required and memory_required must be positive")

        task = Task(
            name=str(data["name"]),
            description=str(data.get("description", "")),
            cpu_required=cpu,
            memory_required=mem,
            priority=priority,
            deadline=data.get("deadline"),
        )
        self._tasks[task.id] = task
        return task

    def get_task(self, task_id: str) -> Task:
        task = self._tasks.get(task_id)
        if task is None:
            raise KeyError(f"Task {task_id} not found")
        return task

    def list_tasks(self, status: str | None = None) -> list[Task]:
        tasks = list(self._tasks.values())
        if status:
            tasks = [t for t in tasks if t.status.value == status.upper()]
        return tasks

    def update_task(self, task_id: str, data: dict) -> Task:
        task = self.get_task(task_id)
        if "name" in data:
            task.name = str(data["name"])
        if "description" in data:
            task.description = str(data["description"])
        if "priority" in data:
            priority = int(data["priority"])
            if not 1 <= priority <= 10:
                raise ValueError("priority must be between 1 and 10")
            task.priority = priority
        if "deadline" in data:
            task.deadline = data["deadline"]
        task.urgency_score = task.calculate_urgency()
        return task

    def cancel_task(self, task_id: str) -> Task:
        task = self.get_task(task_id)
        if task.status in (TaskStatus.COMPLETED, TaskStatus.CANCELLED, TaskStatus.FAILED):
            raise ValueError(f"Task is already {task.status.value}")
        server = self._servers.get(task.server_id) if task.server_id else None
        self._resource_manager.cancel_task(task, server)
        return task

    # ── Server CRUD ────────────────────────────────────────────────────────────

    def create_server(self, data: dict) -> Server:
        required = ["name", "total_cpu", "total_memory"]
        for field in required:
            if field not in data:
                raise ValueError(f"Missing required field: {field}")
        cpu = float(data["total_cpu"])
        mem = float(data["total_memory"])
        if cpu <= 0 or mem <= 0:
            raise ValueError("total_cpu and total_memory must be positive")

        server = Server(
            name=str(data["name"]),
            total_cpu=cpu,
            total_memory=mem,
        )
        self._servers[server.id] = server
        return server

    def get_server(self, server_id: str) -> Server:
        server = self._servers.get(server_id)
        if server is None:
            raise KeyError(f"Server {server_id} not found")
        return server

    def list_servers(self, status: str | None = None) -> list[Server]:
        servers = list(self._servers.values())
        if status:
            servers = [s for s in servers if s.status.value == status.upper()]
        return servers

    def update_server(self, server_id: str, data: dict) -> Server:
        server = self.get_server(server_id)
        if "name" in data:
            server.name = str(data["name"])
        if "status" in data:
            server.status = ServerStatus(data["status"].upper())
        return server

    def remove_server(self, server_id: str) -> Server:
        server = self.get_server(server_id)
        if server.tasks:
            raise ValueError("Cannot remove server with active tasks")
        del self._servers[server_id]
        return server

    # ── Scheduling ─────────────────────────────────────────────────────────────

    def run_scheduler(self, algorithm: str | None = None) -> dict:
        if algorithm:
            self._scheduler.set_algorithm(algorithm)

        pending = [t for t in self._tasks.values() if t.status == TaskStatus.PENDING]
        all_tasks = list(self._tasks.values())
        servers = list(self._servers.values())

        assignments = self._scheduler.schedule_tasks(pending, servers, all_tasks)

        results = []
        for task, server in assignments:
            self._resource_manager.allocate_task(task, server)
            results.append({"task_id": task.id, "server_id": server.id, "task_name": task.name, "server_name": server.name})

        return {
            "algorithm": self._scheduler._algorithm,
            "assignments": results,
            "scheduled_count": len(results),
            "pending_remaining": len(pending) - len(results),
        }

    # ── Stats ──────────────────────────────────────────────────────────────────

    def get_stats(self) -> dict:
        tasks = list(self._tasks.values())
        servers = list(self._servers.values())

        status_counts: dict[str, int] = {}
        for t in tasks:
            status_counts[t.status.value] = status_counts.get(t.status.value, 0) + 1

        sla_at_risk = sum(1 for t in tasks if t.is_sla_at_risk())

        total_cpu = sum(s.total_cpu for s in servers)
        total_mem = sum(s.total_memory for s in servers)
        avail_cpu = sum(s.available_cpu for s in servers)
        avail_mem = sum(s.available_memory for s in servers)

        cpu_util = round((total_cpu - avail_cpu) / total_cpu * 100, 2) if total_cpu > 0 else 0.0
        mem_util = round((total_mem - avail_mem) / total_mem * 100, 2) if total_mem > 0 else 0.0

        return {
            "total_tasks": len(tasks),
            "task_status_counts": status_counts,
            "sla_at_risk": sla_at_risk,
            "total_servers": len(servers),
            "active_servers": sum(1 for s in servers if s.status == ServerStatus.ACTIVE),
            "cluster_cpu_utilization_pct": cpu_util,
            "cluster_memory_utilization_pct": mem_util,
        }
